be1-go/channel/federation/federation.go

```diff
@@ -475,6 +534,100 @@ func (c *Channel) processChallengeRequest(msg message.Message,
 
+func (c *Channel) processFederationResult(msg message.Message,
+	msgData interface{}, s socket.Socket) error {
+	_, ok := msgData.(*messagedata.FederationResult)
+	if !ok {
+		return xerrors.Errorf("message %v is not a federation#result message",
+			msgData)
+	}
+
+	var federationResult messagedata.FederationResult
+
+	err := msg.UnmarshalData(&federationResult)
+	if err != nil {
+		return xerrors.Errorf("failed to unmarshal FederationResult data: %v", err)
+	}
+
+	if federationResult.Status != "success" {
+		if len(federationResult.Reason) > 0 {
+			return xerrors.Errorf("failed to establish federated connection: %v", federationResult.Reason)
+		}
+		return xerrors.Errorf("failed to establish federated connection")
+	}
+
+	var federationChallenge messagedata.FederationChallenge
+	err = federationResult.ChallengeMsg.UnmarshalData(&federationChallenge)
+	if err != nil {
+		return xerrors.Errorf("failed to unmarshal challenge from FederationResult data: %v", err)
+	}
+
+	challengeDataBytes, err := base64.URLEncoding.DecodeString(federationResult.ChallengeMsg.Data)
+	if err != nil {
+		return xerrors.Errorf("failed to decode challenge data in FederationResult: %v", err)
+
+	}
+
+	challengeSignatureBytes, err := base64.URLEncoding.DecodeString(federationResult.ChallengeMsg.Signature)
+	if err != nil {
+		return xerrors.Errorf("failed to decode challenge signature in FederationResult: %v", err)
+
+	}
+
+	remoteOrg := c.getRemoteOrganization(federationResult.ChallengeMsg.Sender)
+	remoteOrg.Lock()
+	defer remoteOrg.Unlock()
+
+	if remoteOrg.state != WaitResult {
+		return answer.NewInternalServerError(invalidStateError, remoteOrg.state, msg)
+
+	}
+
+	pkBytes, err := base64.URLEncoding.DecodeString(remoteOrg.organizerPk)
+	if err != nil {
+		return xerrors.Errorf("failed to decode remote organizers public key: %v", err)
+
+	}
+
+	remotePk := crypto.Suite.Point()
+
+	err = remotePk.UnmarshalBinary(pkBytes)
+	if err != nil {
+		return xerrors.Errorf("failed to decode remote organizers public key: %v", err)
+
+	}
+	err = schnorr.Verify(crypto.Suite, remotePk, challengeDataBytes, challengeSignatureBytes)
+	if err != nil {
+		return xerrors.Errorf("failed to verify signature on challenge in FederationResult message: %v", err)
+
+	}
+
+	resultPkBytes, err := base64.URLEncoding.DecodeString(federationResult.PublicKey)
+	if err != nil {
+		return xerrors.Errorf("failed to decode local public key in FederationResult message: %v", err)
+
+	}
+	localPkBytes, err := c.hub.GetPubKeyOwner().MarshalBinary()
+	if err != nil {
+		return xerrors.Errorf("failed to marshal local organizer public key: %v", err)
+
+	}
+	if !(bytes.Equal(resultPkBytes, localPkBytes)) {
+		return xerrors.Errorf("invalid public key contained in FederationResult message")
+
+	}
+
+	//err = schnorr.Verify(crypto.Suite, remotePk, localPkBinary, pkSignatureBytes)
+	//if err != nil {
+	//	return xerrors.Errorf("failed to verify remote signature on local organizer public key: %v", err)
+
+	//}
```

Comment:
If the PublicKey is only the public key of Alice, then we could directly compare it to c.localOrganizerPk
```suggestion
	if c.localOrganizerPk != federationResult.PublicKey {
		return xerrors.Errorf("invalid public key contained in FederationResult message")
	}
```